src/validation.cpp

```diff
@@ -975,20 +1007,96 @@ bool MemPoolAccept::Finalize(const ATMPArgs& args, Workspace& ws)
 
     // trim mempool and check if tx was trimmed
-    if (!bypass_limits) {
+    // If we are validating a package, don't trim here because we could evict a previous transaction
+    // in the package. LimitMempoolSize() should be called at the very end to make sure the mempool
+    // is still within limits and package submission happens atomically.
+    if (!args.m_package_submission && !bypass_limits) {
         LimitMempoolSize(m_pool, m_active_chainstate.CoinsTip(), gArgs.GetIntArg("-maxmempool", DEFAULT_MAX_MEMPOOL_SIZE) * 1000000, std::chrono::hours{gArgs.GetIntArg("-mempoolexpiry", DEFAULT_MEMPOOL_EXPIRY)});
         if (!m_pool.exists(GenTxid::Txid(hash)))
             return state.Invalid(TxValidationResult::TX_MEMPOOL_POLICY, "mempool full");
     }
     return true;
 }
 
+bool MemPoolAccept::FinalizePackage(const ATMPArgs& args, std::vector<Workspace>& workspaces,
+                                    PackageValidationState& package_state,
+                                    std::map<const uint256, const MempoolAcceptResult>& results)
+{
+    AssertLockHeld(cs_main);
+    AssertLockHeld(m_pool.cs);
+    // The mempool will not be trimmed between individual transaction submissions. Don't allow
+    // memory usage of the mempool to exceed the max by more than MAX_DISCONNECTED_TX_POOL_SIZE.
+    // This is different from the total virtual size of the transactions.
+    const unsigned int total_memory_usage = std::accumulate(workspaces.cbegin(), workspaces.cend(), 0,
+        [](unsigned int sum, const auto& ws) { return sum + ws.m_entry->DynamicMemoryUsage(); });
+    if (total_memory_usage > MAX_DISCONNECTED_TX_POOL_SIZE * 1000) {
+        return package_state.Invalid(PackageValidationResult::PCKG_POLICY, "package memory usage too high");
+    }
+    bool all_submitted = true;
+    // ConsensusScriptChecks adds to the script cache and is therefore consensus-critical;
+    // CheckInputsFromMempoolAndCache asserts that transactions only spend coins available from the
+    // mempool or UTXO set. Submit each transaction to the mempool immediately after calling
+    // ConsensusScriptChecks to make the outputs available for subsequent transactions.
+    for (Workspace& ws : workspaces) {
+        // Since PolicyScriptChecks() passed, this should never fail.
+        if (!ConsensusScriptChecks(args, ws)) {
+            results.emplace(ws.m_ptx->GetWitnessHash(), MempoolAcceptResult::Failure(ws.m_state));
+            all_submitted = package_state.Invalid(PackageValidationResult::PCKG_TX, "transaction failed");
+        }
+
+        // Mempool ancestors may have changed since the last calculation done in PreChecks, since
+        // package ancestors have already been submitted. Since PreChecks() and
+        // PackageMempoolChecks() both enforce ancestor/descendant limits, this should never fail.
+        std::string err_string;
+        if(!m_pool.CalculateMemPoolAncestors(*ws.m_entry, ws.m_ancestors, m_limit_ancestors,
```

Comment:
I've added a comment to explain why we need to do this - `addUnchecked` needs a `setAncestors`. We need to recalculate it in here because the transaction might have new ancestors (i.e. preceding transactions in the package that were submitted).